### PR TITLE
Fix bug with generalized local opens

### PR DESCRIPTION
**Pull request by lpw25**
Since local opens were generalized to allow arbitrary module expressions they are now able to bring fresh types into the environment. This means that they must take care not to include such types in their return type. The current handling is slightly wrong which leads to unnecessary errors like:

```ocaml
module Ext (X : sig type 'a t end) = struct
  type t = T : 'a X.t -> t
end;;

let foo (x : Ext(List).t) =
  match x with
  | T l ->
    let open Ext(Array) in
    T (Array.of_list l);;
[%%expect {|
module Ext :
  functor (X : sig type 'a t end) -> sig type t = T : 'a X.t -> t end
Lines 8-9, characters 4-23:
8 | ....let open Ext(Array) in
9 |     T (Array.of_list l)..
Error: This expression has type t but an expression was expected of type
         Ext(Array).t
       The type constructor t would escape its scope
       t is abstract because no corresponding cmi file was found in path.
|}]
```
This PR fixes this problem by being a bit more careful. The code is now the same as the equivalent code for local module definitions.

### CODE REVIEW

**Review by trefis — APPROVED** [review]
LGTM.

This also makes the error message better in the case where we can't give a type to the returned value, e.g.
```ocaml
let open struct type t = A in A
```
previously this would notice the scope escape, but then some heuristic later one would decide that "oh, that type wasn't present in the environment, let's say that it's probably because of a missing cmi".
Now it will just report the scope escape.

I guess a related subtlety is that you're careful to create `tv` before typing the open declaration which will introduce a new scope; this does indeed work as local modules, but the code is a bit less obvious.

Anyway, all fine.

**Comment by lpw25** [comment]
> I guess a related subtlety is that you're careful to create tv before typing the open declaration which will introduce a new scope; this does indeed work as local modules, but the code is a bit less obvious.

I think that we should seriously consider starting to do the work to make the relationship between levels and the environment explicit. That would allow use to just use the old environment here instead of keeping a unification variable for the purpose. Not in this PR though obviously.

**Comment by trefis** [comment]
Note: this is missing a changes entry. After that we can merge.

**Comment by lpw25** [comment]
I've added a Changes entry in the 4.12 section, since I assume that we're still cherry-picking small bugfixes like this to that branch.

**Comment by trefis** [comment]
It seems reasonable to me, but I'll let @Octachron confirm (and merge if he agrees).

**Comment by Octachron** [comment]
Cherry-picked on 4.12 with 25de00534216172bad3e .

**Pull request merged**
